### PR TITLE
Add cross-compilation support for `rustc`

### PR DESCRIPTION
**Pull request by acfoltzer**
This patch is largely modeled on the relatively-straightforward code for Fortran cross-compilation, so there might be some intricacies missing.

Also, it's hard to follow what's going on with cross-compilation in the test suite (when I try to invoke it, I just get a bunch of `Generating the build system failed.`), so this has only been tested on my actual projects. I did add the appropriate binary to the cross file that's used for the tests, though.

### CODE REVIEW

File: docs/markdown/snippets/rust-cross.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+## Rust cross-compilation
+
+Cross-compilation is now supported for Rust targets. Like other
+cross-compilers, `rustc` must be specified in your cross file.  It
```
Comment (QuLogic) [inline docs/markdown/snippets/rust-cross.md]: The variable is `rust`.

Comment (acfoltzer) [reply docs/markdown/snippets/rust-cross.md]: Good catch, thanks

**Comment by jpakkane** [comment]
The build failure can be fixed by removing the comma and newline from the rust definition cross file. However running that gets me this:

    error[E0463]: can't find crate for `std`
      |
      = note: the `arm-unknown-linux-gnueabihf` target may not be installed

    error: aborting due to previous error

I can't see a package in Ubuntu that would provide arm support for rustc. Do you know what needs to be installed for it to work (if yes, add it to `ciimage/Dockerfile`)?

**Comment by acfoltzer** [comment]
@jpakkane thanks, I had missed that error log.

Regarding the missing target, the usual way would be to use `rustup`:
```
$ rustup target add arm-unknown-linux-gnueabihf
```
However in the `Dockerfile` I don't see how any of the Rust infrastructure is being installed. Is there some other script that sets up `rustup`?

**Comment by jpakkane** [comment]
The container does not user rustup but instead the packages provided by Debian unstable.

**Comment by acfoltzer** [comment]
Gotcha. I can't find where that's happening in https://github.com/mesonbuild/meson/blob/master/ciimage/Dockerfile, but I would be happy to adapt it to use Rustup instead. There's no way to add the cross compilation targets otherwise that wouldn't be extremely brittle.

**Comment by jpakkane** [comment]
I'd like to stick to distro packages in CI as much as possible rather than downloading random versions of stuff from the Internet. This is because the CI platform should be as stable as possible so we don't have to spend time chasing toolchain problems (we have enough random CI breakage as it is already).

If you pinkie swear that you tested this and there is no way to install a cross compiler for Rust using distro packages, we could remove the Rust cross tests from being run on the CI and merge this.

**Comment by acfoltzer** [comment]
I definitely understand the concern around toolchain stability. For what it's worth, stability is a high priority of rustup and the Rust ecosystem in general, so using rustup to install a particular toolchain version would be quite stable. For example, `rustup toolchain install 1.23.0` will always install the same compiler. However the `rustup` executable itself does not have the same stability guarantee, and if everything else is distro packaged, it'd be a pain to have this one exception.

Either way, after the extra commit tweaking the `rust/4 polyglot` test case, I get the following results when running a modified `run_cross_test.py`:

```
$ ./run_cross_test.py cross/ubuntu-armhf.txt 

Running rust tests.

Succeeded test: test cases/rust/1 basic
Succeeded test: test cases/rust/2 sharedlib
Succeeded test: test cases/rust/3 staticlib
Succeeded test: test cases/rust/4 polyglot
Succeeded test: test cases/rust/5 polyglot static
Succeeded test: test cases/rust/6 named staticlib

Total configuration time: 9.16s
Total build time: 7.24s
Total test time: 0.03s

Total passed cross tests: 6
Total failed cross tests: 0
Total skipped cross tests: 0
```
